allow kubectl subcmds to process multiple resources

metral opened the pull request:
~~autoscale, expose & patch~~ Many kubectl subcommands were limited to processing one resource at a time.

This PR allows those subcommands to process multiple resources.

This PR is in reference to https://github.com/kubernetes/kubernetes/pull/23116#issuecomment-202360784 by @deads2k

metral commented:
@smarterclayton @deads2k are we ok to test this?

smarterclayton commented:
ok to test

metral commented:
Anything else needed to merge this PR @smarterclayton ?

metral commented:
Bump. @smarterclayton

smarterclayton commented:
Sorry, been slammed on issues the last few days.  Assigning to @deads2k if he has bandwidth.

metral commented:
@deads2k do you have any availability to review this PR?

deads2k commented:
> @deads2k do you have any availability to review this PR?

smarterclayton and I are working toward the same release.  You're on my list.

deads2k commented:
This adds multiple resource handling for `autoscale`, `expose`, `patch` when targeting a directory.  `autoscale` seems obvious: "scale everything in here to X".  `expose` with no-args is also pretty obvious: "expose all these things (maybe I have a dir full of rcs)".

`patch` is a little less obvious.  The most common usages that apply to multiple objects are handled by other commands already.  I could go either way on this one.

/cc @kubernetes/kubectl  @bgrant0607  for comment

deads2k commented:
Assuming @bgrant0607 approves, you'll need tests in test-cmd

deads2k commented on pkg/kubectl/cmd/autoscale.go:
```diff
@@ -118,62 +110,73 @@ func RunAutoscale(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []
+		info := infos[i]
+		mapping := info.ResourceMapping()
+		if err := f.CanBeAutoscaled(mapping.GroupVersionKind.GroupKind()); err != nil {
+			return err
```
I'd prefer a skip with a message going out to stderr.  If nothing gets scaled, non-zero return code.  If something gets scaled, then a zero return code.

bgrant0607 replied:
Please add that to the conventions:
https://github.com/kubernetes/kubernetes/blob/master/docs/devel/kubectl-conventions.md

deads2k commented on pkg/kubectl/cmd/autoscale.go:
```diff
@@ -118,62 +110,73 @@ func RunAutoscale(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []
+		// Generate new object
+		object, err := generator.Generate(params)
+		if err != nil {
+			return err
```
These errors should be aggregated.  It should result in a non-zero return code, but we should continue to the next referenced resource.

bgrant0607 replied:
Please ensure that's in the conventions, also.

kargakis replied:
https://github.com/kubernetes/kubernetes/blob/4c7abddc1c2c499b58aae44c643235acedee6b10/pkg/util/errors/errors.go#L34

metral replied:
This bit of code was updated in my latest push. I use the `Visit()` to not only traverse the files but because it has error aggregation built in: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource/visitor.go#L202

deads2k commented on pkg/kubectl/cmd/autoscale.go:
```diff
@@ -118,62 +110,73 @@ func RunAutoscale(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []
-	if err := kubectl.EnsureFlagsValid(cmd, generators, generatorName); err != nil {
-		return err
-	}
+		name := info.Name
```
From here to the end of the loop feels like it should be factored into a different method.

metral replied:
Extracting this section of code into its own method would result in essentially duplicating the `RunAutoscale` func, as each iteration relies on many dependencies set up through out the entire func.

A refactor appears like it would be best suited in a larger, separate PR.

Thoughts?

deads2k replied:
> A refactor appears like it would be best suited in a larger, separate PR.
> 
> Thoughts?

Worrisome, but I'm ok with a followup issue linked back here.

metral replied:
Sounds good - I'll create one

deads2k commented on pkg/kubectl/cmd/expose.go:
```diff
@@ -149,101 +137,124 @@ func RunExpose(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
-		name = name[:validation.DNS952LabelMaxLength]
+
+	// Get the input object
+	inputObject, err := r.Object()
```
Why can't I just use the `infos`?

deads2k commented on pkg/kubectl/cmd/expose.go:
```diff
@@ -149,101 +137,124 @@ func RunExpose(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
+		// For objects that need a pod selector, derive it from the exposed object in case a user
+		// didn't explicitly specify one via --selector
+		if s, found := params["selector"]; found && kubectl.IsZero(s) {
+			s, err := f.MapBasedSelectorForObject(inputObjects[i])
```
sometimes using `infos` and sometimes using `inputObjects` seems strange.  What's going on that requires this?

metral replied:
good catch. I'll fix this up

deads2k commented:
@metral I'd wait for consensus on the idea before spending the time updating the pull.  

I like the changes.

bgrant0607 commented on pkg/kubectl/cmd/autoscale.go:
```diff
@@ -118,62 +110,73 @@ func RunAutoscale(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []
-	params["scaleRef-kind"] = mapping.GroupVersionKind.Kind
-	params["scaleRef-name"] = name
-	params["scaleRef-apiVersion"] = mapping.GroupVersionKind.GroupVersion().String()
+	for i := range infos {
```
Why not use Visit?

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/create.go#L121

metral replied:
`infos` is the result of `r.Infos()`, where the `Infos()` method makes use of `Visit()`: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/resource/result.go#L94-L114

Am I missing something?

bgrant0607 commented:
Yes, I'm in favor of supporting multiple resources in all commands where it could plausibly be useful, including all of these.

metral committed:
allow kubectl subcmds to process multiple resources

- use resource.Visit() to recursively process resources, as well as, aggregate
errors where possible

metral commented:
@deads2k I've gone ahead and addressed all feedback including: enabling recursive functionality in the subcommands missing this ability, as well as, aggregating any errors when working recursively. I've also added all the appropriate tests for the subcommands around recursiveness + error aggregation in `hack/test-cmd.sh`.

All of the PR checks are passing. Is there anything else needed for this PR?

metral commented:
@deads2k Do you happen to have an estimate as to when you can get to this PR? Thanks!

deads2k commented on pkg/kubectl/cmd/get.go:
```diff
@@ -262,7 +275,8 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
 			objs[ix], err = infos[ix].Mapping.ConvertToVersion(infos[ix].Object, version.String())
 			if err != nil {
-				return err
+				allErrs = append(allErrs, err)
```
This command changed to update error handling?

metral replied:
Yes, that is correct. `kubectl get` was refactored to use `Visit()` rather than `Infos()`, so that we can aggregate the errors when processing a dir recursively.

deads2k commented on pkg/kubectl/cmd/rollout/rollout_pause.go:
```diff
@@ -75,7 +73,7 @@ func NewCmdRolloutPause(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 }
 
-func (o *PauseConfig) CompletePause(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, args []string) error {
+func (o *PauseConfig) RunPause(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, args []string) error {
```
`run` methods shouldn't require commands or factories.  Rather than hold up the rest of the work on this, splitting it out is fine.

kargakis replied:
See https://github.com/kubernetes/kubernetes/blob/master/docs/devel/kubectl-conventions.md#command-implementation-conventions

deads2k commented on pkg/kubectl/cmd/rollout/rollout_resume.go:
```diff
@@ -73,7 +71,7 @@ func NewCmdRolloutResume(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 }
 
-func (o *ResumeConfig) CompleteResume(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, args []string) error {
+func (o *ResumeConfig) RunResume(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, args []string) error {
```
same comments on run here.

deads2k commented on pkg/kubectl/cmd/rollout/rollout_undo.go:
```diff
@@ -73,7 +71,7 @@ func NewCmdRolloutUndo(f *cmdutil.Factory, out io.Writer) *cobra.Command {
 }
 
-func (o *UndoOptions) CompleteUndo(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, args []string) error {
+func (o *UndoOptions) RunUndo(f *cmdutil.Factory, cmd *cobra.Command, out io.Writer, args []string) error {
```
same

deads2k commented on pkg/kubectl/cmd/scale.go:
```diff
@@ -122,43 +121,35 @@ func RunScale(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []stri
 
-	resourceVersion := cmdutil.GetFlagString(cmd, "resource-version")
-	if len(resourceVersion) != 0 && len(infos) > 1 {
-		return fmt.Errorf("cannot use --resource-version with multiple resources")
```
I think you still need this check

metral replied:
@deads2k I had removed the checks for the variables `resourceVersion` and for `currentSize` (below) because as we traverse each file using `Visit()` now, they're both being scoped to a single item, not multiple, and it is still allowing the precondition to be validated.

Do you still think these checks are still necessary given that the usage of `infos` is being superseded by using `Visit()`?

deads2k replied:
> Do you still think these checks are still necessary given that the usage of infos is being superseded by using Visit()?

If someone specifies `--resourceversion` for multiple objects, you guaranteed that it cannot succeed on more than one of them.

metral replied:
Ah, true. Ok thanks

On Thursday, April 28, 2016, David Eads <notifications@github.com> wrote:

> In pkg/kubectl/cmd/scale.go
> <https://github.com/kubernetes/kubernetes/pull/23673#discussion_r61475083>
> :
>
> >
> > -	resourceVersion := cmdutil.GetFlagString(cmd, "resource-version")
> > -	if len(resourceVersion) != 0 && len(infos) > 1 {
> > -		return fmt.Errorf("cannot use --resource-version with multiple resources")
>
> Do you still think these checks are still necessary given that the usage
> of infos is being superseded by using Visit()?
>
> If someone specifies --resourceversion for multiple objects, you
> guaranteed that it cannot succeed on more than one of them.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23673/files/c04de61a6caf10487672f05b72a1c81b5a83061d#r61475083>
>


-- 
-Mike Metral

deads2k commented on pkg/kubectl/cmd/scale.go:
```diff
@@ -122,43 +121,35 @@ func RunScale(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []stri
-	}
-	currentSize := cmdutil.GetFlagInt(cmd, "current-replicas")
-	if currentSize != -1 && len(infos) > 1 {
-		return fmt.Errorf("cannot use --current-replicas with multiple resources")
```
and probably this one.

deads2k replied:
> and probably this one.

This one I could probably live with removing.  It's a little strange, but its not doomed to failure.

deads2k commented on pkg/kubectl/cmd/edit.go:
```diff
@@ -167,24 +169,32 @@ func RunEdit(f *cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args
 	defaultVersion, err := cmdutil.OutputVersion(cmd, clientConfig.GroupVersion)
 	if err != nil {
-		return err
+		allErrs = append(allErrs, err)
```
I think this error is still fatal.

deads2k commented on pkg/kubectl/cmd/edit.go:
```diff
@@ -159,6 +160,7 @@ func RunEdit(f *cmdutil.Factory, out, errOut io.Writer, cmd *cobra.Command, args
 		NamespaceParam(cmdNamespace).DefaultNamespace().
 		FilenameParam(enforceNamespace, options.Recursive, options.Filenames...).
 		ResourceTypeOrNameArgs(true, args...).
+		ContinueOnError().
```
I'd suggest splitting this part out of this pull.  `edit` is complicated enough that it would take a long time to review and I wouldn't want to hold the rest of the work on this.

kargakis replied:
agreed, let's leave edit out for now

deads2k commented:
I like the pull and I like the tests.  I'd like to slim it down slightly and merge most of it, then come back to the really sticky bits.

kargakis commented on hack/testdata/recursive/deployment/busybox.yaml:
```diff
@@ -0,0 +1,18 @@
+    spec:
+      containers:
+      - name: nginx
+        image: gcr.io/google-containers/nginx:1.7.9
```
this is not busybox

kargakis commented on hack/testdata/recursive/deployment/deployment/busybox-broken.yaml:
```diff
@@ -0,0 +1,18 @@
+    spec:
+      containers:
+      - name: nginx
+        image: gcr.io/google-containers/nginx:1.7.9
```
not buxybox. also how is this deployment broken?

metral replied:
the `kind` key has an intentional typo to issue a simple error upon usage with the subcommands

kargakis replied:
ok

kargakis commented on hack/testdata/recursive/deployment/deployment/busybox.yaml:
```diff
@@ -0,0 +1,18 @@
+    spec:
+      containers:
+      - name: nginx
+        image: gcr.io/google-containers/nginx:1.7.9
```
not busybox

metral replied:
good catch. i'll update all filenames

metral commented:
@deads2k I've gone ahead and addressed the latest feedback

metral commented:
@deads2k anything else needed?

deads2k commented:
lgtm, squash down.

metral commented:
The Jenkins unit/integration test is failing because in my addition of the `rollout history` test to `hack/test-cmd.sh`, `rollout history` is hitting a `nil pointer dereference`. see issue: https://github.com/kubernetes/kubernetes/issues/24606

metral commented:
@deads2k due to the error encountered in my comment above, I pulled out the `rollout history` bits & its test to move this PR along. It just needs an LGTM
